web-common/src/main/java/org/jboss/as/web/session/CompositeCharSequence.java

```diff
@@ -0,0 +1,94 @@
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.jboss.as.web.session;
+
+import java.nio.CharBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * A {@link CharSequence} composed of other character sequences.
+ * All methods delegate to one or more of the underlying character sequences, using relative indexes.
+ * @author Paul Ferraro
+ */
+public class CompositeCharSequence implements CharSequence {
+    private final List<CharSequence> sequences;
+
+    public CompositeCharSequence(CharSequence... sequences) {
+        this(Arrays.asList(sequences));
+    }
+
+    public CompositeCharSequence(List<CharSequence> sequences) {
+        this.sequences = sequences;
+    }
+
+    @Override
+    public int length() {
+        int length = 0;
+        for (CharSequence sequence : this.sequences) {
+            length += sequence.length();
+        }
+        return length;
+    }
+
+    @Override
+    public char charAt(int index) {
+        int relativeIndex = index;
+        for (CharSequence sequence : this.sequences) {
+            if (relativeIndex < sequence.length()) {
+                return sequence.charAt(relativeIndex);
+            }
+            relativeIndex -= sequence.length();
+        }
+        throw new IndexOutOfBoundsException();
+    }
+
+    @Override
+    public CharSequence subSequence(int start, int end) {
+        if ((start < 0) || (start > end) || (end > this.length())) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (start == end) return "";
+        List<CharSequence> result = new ArrayList<>(this.sequences.size());
+        int relativeStart = start;
+        int relativeEnd = end;
+        for (CharSequence sequence : this.sequences) {
+            if ((relativeStart < sequence.length()) && (relativeEnd > 0)) {
+                result.add(sequence.subSequence(Math.max(relativeStart, 0), Math.min(relativeEnd, sequence.length())));
+            }
+            relativeStart -= sequence.length();
+            relativeEnd -= sequence.length();
+        }
+        return (result.size() == 1) ? result.get(0) : new CompositeCharSequence(result);
+    }
+
+    @Override
+    public String toString() {
+        CharBuffer buffer = CharBuffer.allocate(this.length());
+        for (CharSequence sequence : this.sequences) {
+            buffer.put(CharBuffer.wrap(sequence));
+        }
+        return String.valueOf(buffer.array());
```

Comment:
Is there value in caching this result, at the cost of a pointer per object, to save repeated char[] allocations at L88?  I see a lot of CharSequence.toString() calls in here but I'm not sure in how many the CharSequence is one of these.